### PR TITLE
add TargetHostName to SslStream

### LINKED ISSUE #27619: Add SslStream.ServerName and other properties from SslClientAuthenticationOptions
I would like to use `SocketsHttpHandler` instead of `HttpClientHandler`, because of the additional timeout options. In my current implementation I use `HttpClientHandler.ServerCertificateCustomValidationCallback` to implement custom SSL certificate validation. Inside the callback I use `HttpRequestMessage.RequestUri.Host` to see if the user has configured a trusted certificate for the specified host. If the certificate is not trusted, then I add the certificate to HttpRequestMessage.Properties, so that the requester/caller can show the certificate to the user. When I try to re-implement this using `SocketsHttpHandler.SslOptions.RemoteCertificateValidationCallback`, then I get an `SslStream` instead of a `HttpResponseMessage`. As a result I do not know for which host I am verifying the certificate and I cannot add the certificate to the request properties.

As a workaround I use reflection to get `HttpClientHandler._socketsHttpHandler`, so that I can configure the timeout options of `SocketsHttpHandler`, while still using `HttpClientHandler`. Of course I would prefer not having to resort to using reflection and exploiting implementation details.

Another way to allow this is to expose `System.Net.Http.ConnectHelper.CertificateCallbackMapper`, which is what `HttpClientHandler.ServerCertificateCustomValidationCallback` uses internally.

### PR DESCRIPTION
**Pull request by wfurt**
This exposes internal property used with SNI. This will return string.Empty if not available to be consistent with ServerCertificateSelectionCallback. 

Strangely, when empty name is provided, we fabricate random name instead of skipping SNI extension. This is probably not common case so I made the property to return the fabricated name instead of empty value from SslClientAuthenticationOptions. On server, this gets name requested by client. Since the behavior above, it is not easy to construct test without SNI.

related to #37933
fixes #27619

### CODE REVIEW

**Comment by stephentoub** [comment]
> Strangely, when empty name is provided, we fabricate random name instead of skipping SNI extension

What kind of name do we fabricate?  Can you provide an example?

**Comment by wfurt** [comment]
> > Strangely, when empty name is provided, we fabricate random name instead of skipping SNI extension
> 
> What kind of name do we fabricate? Can you provide an example?

Something like `?124` which looks lame to me but I don't know the history of it.

https://github.com/dotnet/runtime/blob/3495f1ae2d3eea3a98c5bc59f5c61c537348f2cb/src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs#L70-L73

File: src/libraries/System.Net.Security/src/System/Net/Security/SslAuthenticationOptions.cs [hunk 1]
```diff
@@ -21,7 +21,7 @@ internal SslAuthenticationOptions(SslClientAuthenticationOptions sslClientAuthen
             EncryptionPolicy = sslClientAuthenticationOptions.EncryptionPolicy;
             IsServer = false;
             RemoteCertRequired = true;
-            TargetHost = sslClientAuthenticationOptions.TargetHost;
+            TargetHost = sslClientAuthenticationOptions.TargetHost!;
```
[the remark this answers is not in the document]
Comment (wfurt) [reply src/libraries/System.Net.Security/src/System/Net/Security/SslAuthenticationOptions.cs:24]: This is internal copy. The client side is checked for null and server side always assigns string.Empty so I cannot be null as far as I can tell.

**Comment by stephentoub** [comment]
> Something like ?124 which looks lame to me but I don't know the history of it.

That does look really strange.  This was inherited from .NET Framework:
https://referencesource.microsoft.com/#System/net/System/Net/SecureProtocols/_SslState.cs,172
@davidsh, do you know what the purpose of this is?

**Comment by wfurt** [comment]
When no name is provided then it is harder for ssl client to verify serve's name.
But that can be handled in the callback. I did verify that the fabricated name get sent on wire as part of the ClientHello. 
I did not want to make functional change in this PR but I can perhaps open new issue for investigation. I'm wondering if this was/is some limitatuon of Windows Schannel API.

**Review by stephentoub — APPROVED** [review]
We can follow-up on the strange constructed names subsequently.  Is there an issue for that?

File: src/libraries/System.Net.Security/tests/FunctionalTests/SslStreamNetworkStreamTest.cs [hunk 2]
```diff
@@ -193,6 +193,70 @@ public async Task SslStream_NestedAuth_Throws()
             }
         }
 
+        [Theory]
+        [InlineData(false)]
+        [InlineData(true)]
+        public async Task SslStreag_TargetHostName_Succeeds(bool useEmptyName)
```
Comment (stephentoub) [inline src/libraries/System.Net.Security/tests/FunctionalTests/SslStreamNetworkStreamTest.cs:199]:
```suggestion
        public async Task SslStream_TargetHostName_Succeeds(bool useEmptyName)
```

File: src/libraries/System.Net.Security/tests/FunctionalTests/SslStreamNetworkStreamTest.cs [hunk 3]
```diff
@@ -193,6 +193,70 @@ public async Task SslStream_NestedAuth_Throws()
...
+        {
+            string tagetName = useEmptyName ? string.Empty : Guid.NewGuid().ToString("N");
+
+            (Stream clientStream, Stream serverStream) = TestHelper.GetConnectedStreams();
+            using (clientStream)
+            using (serverStream)
+            using (var client = new SslStream(clientStream))
+            using (var server = new SslStream(serverStream))
+            using (X509Certificate2 certificate = Configuration.Certificates.GetServerCertificate())
+            {
```
Comment (stephentoub) [inline src/libraries/System.Net.Security/tests/FunctionalTests/SslStreamNetworkStreamTest.cs:209]: Can you assert here that client.TargetHostName and server.TargetHostName are not null?

**Comment by wfurt** [comment]
> We can follow-up on the strange constructed names subsequently. Is there an issue for that?

 #38356

File: src/libraries/System.Net.Security/src/System/Net/Security/SslAuthenticationOptions.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (wfurt) [reply src/libraries/System.Net.Security/src/System/Net/Security/SslAuthenticationOptions.cs:27]: done.

File: src/libraries/System.Net.Security/tests/FunctionalTests/SslStreamNetworkStreamTest.cs (hunk 3 above)
Comment (wfurt) [reply src/libraries/System.Net.Security/tests/FunctionalTests/SslStreamNetworkStreamTest.cs:209]: updated.

**Pull request merged**